staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/maplist.go

```diff
@@ -17,162 +17,13 @@ limitations under the License.
-// keys.
-func (ks *multiKeyStrategy) CompositeKeyFor(obj map[string]interface{}) (interface{}, bool) {
-	const keyDelimiter = "\x00" // 0 byte should never appear in the composite key except as delimiter
-
-	var delimited strings.Builder
-	for _, key := range ks.sts.XListMapKeys {
-		v, ok := obj[key]
-		if !ok {
-			return nil, false
-		}
-
-		switch v.(type) {
-		case bool:
-			fmt.Fprintf(&delimited, keyDelimiter+"%t", v)
-		case float64:
-			fmt.Fprintf(&delimited, keyDelimiter+"%f", v)
-		case int64:
-			fmt.Fprintf(&delimited, keyDelimiter+"%d", v)
-		case string:
-			fmt.Fprintf(&delimited, keyDelimiter+"%q", v)
-		default:
-			return nil, false // values must be scalars
-		}
-	}
-	return delimited.String(), true
-}
-
-// emptyMapList is a mapList containing no elements.
-type emptyMapList struct{}
-
-func (emptyMapList) get(interface{}) interface{} {
-	return nil
-}
-
-type mapListImpl struct {
-	sts *schema.Structural
-	ks  keyStrategy
-	// keyedItems contains all lazily keyed map items
-	keyedItems map[interface{}]interface{}
-	// unkeyedItems contains all map items that have not yet been keyed
-	unkeyedItems []interface{}
-}
-
-func (a *mapListImpl) get(obj interface{}) interface{} {
-	mobj, ok := obj.(map[string]interface{})
-	if !ok {
-		return nil
-	}
-
-	key, ok := a.ks.CompositeKeyFor(mobj)
-	if !ok {
-		return nil
-	}
-	if match, ok := a.keyedItems[key]; ok {
-		return match
-	}
-	// keep keying items until we either find a match or run out of unkeyed items
-	for len(a.unkeyedItems) > 0 {
-		// dequeue an unkeyed item
-		item := a.unkeyedItems[0]
-		a.unkeyedItems = a.unkeyedItems[1:]
-
-		// key the item
-		mitem, ok := item.(map[string]interface{})
-		if !ok {
-			continue
-		}
-		itemKey, ok := a.ks.CompositeKeyFor(mitem)
-		if !ok {
-			continue
-		}
-		if _, exists := a.keyedItems[itemKey]; !exists {
-			a.keyedItems[itemKey] = mitem
-		}
-
-		// if it matches, short-circuit
-		if itemKey == key {
-			return mitem
-		}
-	}
-
-	return nil
-}
-
-func makeKeyStrategy(sts *schema.Structural) keyStrategy {
-	if len(sts.XListMapKeys) == 1 {
-		key := sts.XListMapKeys[0]
-		return &singleKeyStrategy{
-			key: key,
-		}
-	}
-
-	return &multiKeyStrategy{
-		sts: sts,
-	}
-}
-
 // makeMapList returns a queryable interface over the provided x-kubernetes-list-type=map
 // keyedItems. If the provided schema is _not_ an array with x-kubernetes-list-type=map, returns an
 // empty mapList.
-func makeMapList(sts *schema.Structural, items []interface{}) (rv mapList) {
-	if sts.Type != "array" || sts.XListType == nil || *sts.XListType != "map" || len(sts.XListMapKeys) == 0 || len(items) == 0 {
-		return emptyMapList{}
-	}
-	ks := makeKeyStrategy(sts)
-	return &mapListImpl{
-		sts:          sts,
-		ks:           ks,
-		keyedItems:   map[interface{}]interface{}{},
-		unkeyedItems: items,
-	}
+func makeMapList(sts *schema.Structural, items []interface{}) (rv celopenapi.MapList) {
+	return celopenapi.MakeMapList(sts.ToKubeOpenAPI(), items)
```

Comment:
`sts.ToKubeOpenAPI()` is relatively expensive, right? (it looks like it recursively converts the whole subschema to openapi). There are two issues with that:

1. it looks like this is called on every validate of every custom resource and never cached... @jpbetz, I was expecting any schema/structural conversions to be done once per validator instance (either lazily or at validator construction time), not per item validation
2. do we re-call this multiple times at multiple levels? if so, are we reintroducing a performance problem with deeply nested schemas recursing the same trees multiple times?

Reply:
The runtime cost of `ToKubeOpenAPI()` depends only on the size of the input CRD. There is no "subschema" because CRD cannot embed other types. Preserve Unknown Fields does not increase the depth of the schema either.

1. The validator is indeed cached. It is stored as part of the storage strategy.
https://github.com/kubernetes/kubernetes/blob/v1.26.0/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L52

2. Yes deeply nested schemas are a problem. The worst case scenario would be 
```yaml
    openAPIV3Schema:
      type: object
      properties:
        spec:
          type: object
          x-kubernetes-validations: # ...
          properties:
            nested1:
              type: object
	      x-kubernetes-validations: # ...
	      properties:
	        nested2:
	        type: object
	        x-kubernetes-validations: # ...
	        # keep nesting
``` 
Because of the conversion each time, the time complexity degrades into O(N ^ 2) from O(N)

Reply:
> The validator is indeed cached. It is stored as part of the storage strategy.

The validator is cached, but this wasn't previously being cached along with the validator, right?

Reply:
Yes you are correct. This part was not cached. It would be costly for the conversion to happen every time. However, the saving grace is that the schema for CRD validator comes with the CR being validated. I am working on a conversion-free solution for the "common ground" between CRD Validation and ValidatingAdmissionPolicy features.